Java EE Quickstart article for webapp login

jimmyjames opened the pull request:
New Quickstart article demonstrating web app login using Java EE and Auth0.

🗒 **Notes:**
- The sample that accompanies this article is [here](https://github.com/auth0-samples/auth0-java-ee-sample).

jimmyjames committed:
Java EE Quickstart article for webapp login

jimmyjames committed:
Add logo_name to index.yml to fix broken image (also requires docs platform change)

damieng commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+The Java EE 8 Security API introduced the `HttpAuthenticationMechanism` interface to enable applications to obtain a user's credentials. Default implementations exist for Basic and form-based authentication, and it provides an easy way to configure a custom authentication strategy.
+
+To authenticate users with Auth0, you can implement the `HttpAuthenticationMechanism` interface and some of its collaborators. The key interfaces involved are:
```
Collaborators seems an odd word choice and made me immediately think of human collaborators on a project...

jimmyjames replied:
Yeah, that's not what we want readers to think :smile:

How about, "To authenticate users with Auth0, you can provide custom implementations of the following interfaces:"

jimmyjames replied:
Updated.

jimmyjames committed:
Simplify and don't use the word collaborators, per review

lbalmaceda commented on articles/quickstart/webapp/java-ee/01-login.md:
```diff
@@ -0,0 +1,21 @@
+  - quickstarts
+  - webapp
+  - login
+  - java ee
```
would this work without quotes? or maybe is better `javaee`?

jimmyjames replied:
changed to `java-ee`

lbalmaceda commented on articles/quickstart/webapp/java-ee/_includes/_setup.md:
```diff
@@ -0,0 +1,83 @@
+<env-entry>
+    <env-entry-name>auth0.clientSecret</env-entry-name>
+    <env-entry-type>java.lang.String</env-entry-type>
+    <env-entry-value>YOUR_CLIENT_SECRET</env-entry-value>
```
Are you sure `YOUR_CLIENT_SECRET` is auto-replaced? I recall a `${account.clientSecret}` or similar placeholder that could be used here.

jimmyjames replied:
We don't show secrets in the quickstart articles, so it would display as `YOUR_CLIENT_SECRET`. I will check that the sample code is configured correctly, and if not will make a PR to the sample.

jimmyjames replied:
sample code is configured just as others that work 👍

lbalmaceda commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+    }
+
+    private boolean isCallbackRequest(HttpServletRequest request) {
+        return (request.getRequestURI().equals("/callback") && request.getParameter("code") != null);
```
wrapping `(` `)` could be removed, if desired.

jimmyjames replied:
removed.

lbalmaceda commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+This sample uses the [Authorization Code Flow](https://auth0.com/docs/flows/concepts/auth-code) to exchange an Authorization Code for a token during the authentication flow. If this request is to the `/callback` endpoint and contains the `code` request parameter, it does a few important things:
+
+- Calls the `handle` method of the `AuthenticationController` to exchange the Authorization Code for an ID and access token.
```
Refer to them as separate tokens. 
```suggestion
- Calls the `handle` method of the `AuthenticationController` to exchange the Authorization Code for an ID token and an access token.
```

lbalmaceda commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+The `LoginController` is responsible for redirecting the request to the proper authorization URL, where the user can authenticate with Auth0. It uses the `AuthenticationController` from the Auth0 Java MVC SDK to build the correct authorization URL, using the configuration values injected via `Auth0AuthenticationConfig`. By default, this sample requests the `"openid profile email"` scopes, to allo ...
+
+Once the user has entered their credentials and authorized the requested permissions, Auth0 will issue a request to the `callbackUrl`, and include a `code` query parameter which can be exchanged for an ID and access token. Recall that the `Auth0HttpAuthenticationMechanism` created above handles this exchange so that it can notify the application container of authentication status. This allows the ...
```
Refer to them as separate tokens. 
```suggestion
Once the user has entered their credentials and authorized the requested permissions, Auth0 will issue a request to the `callbackUrl`, and include a `code` query parameter which can be exchanged for an ID token and an access token. Recall that the `Auth0HttpAuthenticationMechanism` created above handles this exchange so that it can notify the application container of authentication status. This allows the Servlet that handles requests to the `/callback` path to simply forward the request on to the originally requested resource prior to logging in, or simply redirect to the home page:
```

lbalmaceda commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+To log a user out, you should clear the application session and log the user out of Auth0. This is handled in the `LogoutServlet`.
+
+First, clear the session by calling `request.getSession().invalidate()`. Then construct the logout URL, being sure to include the `returnTo` query parameter, which is where the user will be redirected to after logging out. Finally, redirect the response to our logout URL:
```
```suggestion
First, clear the session by calling `request.getSession().invalidate()`. Then construct the logout URL, being sure to include the `returnTo` query parameter, which is where the user will be redirected to after logging out. Finally, redirect the response to the application's logout URL:
```

lbalmaceda commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+        }
+    }
+
+    private String getReturnUrl(HttpServletRequest request) {
```
can this be renamed? could be confused with the "return to url" required by the logout endpoint. Maybe "getLogoutUrl" ?

lbalmaceda replied:
Probably affects the sample project code as well

jimmyjames replied:
renamed as suggested. Will update the sample as well. 👍

jimmyjames replied:
update to sample: https://github.com/auth0-samples/auth0-java-ee-sample/pull/13

jimmyjames committed:
alias change and code snippet changes based on review

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_login.md

Co-Authored-By: Luciano Balmaceda <balmacedaluciano@gmail.com>

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_login.md

Co-Authored-By: Luciano Balmaceda <balmacedaluciano@gmail.com>

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_login.md

Co-Authored-By: Luciano Balmaceda <balmacedaluciano@gmail.com>

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_setup.md:
```diff
@@ -0,0 +1,83 @@
+## Configure Java EE to use Auth0
+
+### Setup dependencies
```
```suggestion
### Set up dependencies
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+The Java EE 8 Security API introduced the `HttpAuthenticationMechanism` interface to enable applications to obtain a user's credentials. Default implementations exist for Basic and form-based authentication, and it provides an easy way to configure a custom authentication strategy.
+
+To authenticate users with Auth0, you can provide custom implementations of the following interfaces:
```
Does the user have to provide all these? It sounds like some of these are optional.

jimmyjames replied:
They will typically all need to be implemented, and in the context of this specific quickstart, are all required. How about `To authenticate with Auth0, provide custom implementations of the following interfaces:` (removing `you can`).

khorne3 replied:
Yup, that would've been my exact suggestion

jimmyjames replied:
👍 updated.

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+- `CallerPrincipal`: Represents the caller principal of the current HTTP request ([JavaDoc](https://javaee.github.io/javaee-spec/javadocs/javax/security/enterprise/CallerPrincipal.html)).
+- `Credential`: Represents the credential the caller will use to authenticate ([JavaDoc](https://javaee.github.io/javaee-spec/javadocs/javax/security/enterprise/credential/Credential.html)).
+
+First, make your Auth0 settings available to the application. Create an `@ApplicationScoped` bean to retrieve the values from the web context and make them available via getters:
```
```suggestion
First, make your Auth0 settings available to the application by creating an `@ApplicationScoped` bean to retrieve the values from the web context and make them available via getters:
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
````diff
@@ -0,0 +1,379 @@
+}
+```
+
+Next, create a custom `CallerPrincipal` that represents the caller of the current request.
````
```suggestion
Next, create a custom `CallerPrincipal` that represents the caller of the current request:
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
````diff
@@ -0,0 +1,379 @@
+}
+```
+
+You can now implement a custom `Credential` that will be used to represent the user's credentials. It will hold information about the principal:
````
Would it be appropriate to replace principal (the last word) with `callerPrinciple` to be clearer?

jimmyjames replied:
I don't think so in this case, since text in an inline code block would indicate a specific field/class/etc. For Java developers, the concept of a principal is fairly well-understood and frequently used.

khorne3 replied:
Sounds good -- let's leave it

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
````diff
@@ -0,0 +1,379 @@
+}
+```
+
+If the `credential` is an `Auth0Credential`, the calling user is authenticated and valid, so a `CredentialValidationResult` created with the credential is returned to indicate success. If it is not an `Auth0Credential`, return `CredentialValidationResult.NOT_VALIDATED_RESULT`.
````
Is the user responsible for setting the code up to create `CredentialValidationResult` and return the credential? If so, maybe rewrite this sentence as:

> If the `credential` is an `Auth0Credential`, the calling user is authenticated and valid. Create `CredentialValidationResult` and return it with the credential to indicate success.

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+If the `credential` is an `Auth0Credential`, the calling user is authenticated and valid, so a `CredentialValidationResult` created with the credential is returned to indicate success. If it is not an `Auth0Credential`, return `CredentialValidationResult.NOT_VALIDATED_RESULT`.
+
+Before implementing the `HttpAuthenticationMechanism` interface that will use all these collaborators, create a bean that will provide a configured instance of the `AuthenticationController` from the Auth0 Java MVC SDK, used to build the authorize URLs and handle the token exchange to authenticate users:
```
Maybe something like

```suggestion
Create a bean that will provide a configured instance of the `AuthenticationController` from the Auth0 Java MVC SDK. The bean will be used to build the authorize URLs and handle the token exchange to authenticate users:
```

While reading, it was a bit jarring to see mention to the `HttpAuthenticationMechanism` interface so suddenly.

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
````diff
@@ -0,0 +1,379 @@
+}
+```
+
+Finally, implement a custom `HttpAuthenticationMechanism`
````
```suggestion
Finally, implement a custom `HttpAuthenticationMechanism`  interface that will use all these collaborators
```

I'd also recommend using something more specific instead of `these collaborators` (unless this is something that's immediately obvious to Java devs)

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
````diff
@@ -0,0 +1,379 @@
+}
+```
+
+The class overrides the `validateRequest` method, which is called on every request to our application, and is responsible for notifying the container of the authentication status.
````
```suggestion
The `Auth0AuthenticationMechanism` class overrides the `validateRequest` method, which is called on every request to our application and is responsible for notifying the container of the authentication status.
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+The class overrides the `validateRequest` method, which is called on every request to our application, and is responsible for notifying the container of the authentication status.
+
+This sample uses the [Authorization Code Flow](https://auth0.com/docs/flows/concepts/auth-code) to exchange an Authorization Code for a token during the authentication flow. If this request is to the `/callback` endpoint and contains the `code` request parameter, it does a few important things:
```
I'm struggling a bit to follow the transition from the instructions and code samples above to this section. I don't know if it needs another header, but I think a sentence or two explaining where you're going in the following section would help orient the user

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
````diff
@@ -0,0 +1,379 @@
+}
+```
+
+The `LoginController` is responsible for redirecting the request to the proper authorization URL, where the user can authenticate with Auth0. It uses the `AuthenticationController` from the Auth0 Java MVC SDK to build the correct authorization URL, using the configuration values injected via `Auth0AuthenticationConfig`. By default, this sample requests the `"openid profile email"` scopes, to allo ...
````
```suggestion
The `LoginController` is responsible for redirecting the request to the proper authorization URL, where the user can authenticate with Auth0. It uses the `AuthenticationController` from the Auth0 Java MVC SDK to build the correct authorization URL. This uses the configuration values injected via `Auth0AuthenticationConfig`. By default, the sample requests the `"openid profile email"` scopes to allow the application to retrieve basic profile information from the authenticated user. You can read more about these scopes in the [OpenID Connect Scopes](https://auth0.com/docs/scopes/current/oidc-scopes) documentation.
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+The `LoginController` is responsible for redirecting the request to the proper authorization URL, where the user can authenticate with Auth0. It uses the `AuthenticationController` from the Auth0 Java MVC SDK to build the correct authorization URL, using the configuration values injected via `Auth0AuthenticationConfig`. By default, this sample requests the `"openid profile email"` scopes, to allo ...
+
+Once the user has entered their credentials and authorized the requested permissions, Auth0 will issue a request to the `callbackUrl`, and include a `code` query parameter which can be exchanged for an ID token and an access token. Recall that the `Auth0HttpAuthenticationMechanism` created above handles this exchange so that it can notify the application container of authentication status. This a ...
```
```suggestion
Once the user has entered their credentials and authorized the requested permissions, Auth0 will issue a request to the `callbackUrl` and include a `code` query parameter which can be exchanged for an ID token and an access token. 

Recall that the `Auth0HttpAuthenticationMechanism` created above handles this exchange so that it can notify the application container of authentication status. This allows the Servlet that handles requests to the `/callback` path to forward the request on to the resource originally requested prior to logging in or redirect to the home page:
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+## Handle Logout
+
+To log a user out, you should clear the application session and log the user out of Auth0. This is handled in the `LogoutServlet`.
```
```suggestion
To log a user out of your application, you should clear the application session and log the user out of Auth0. This is handled in `LogoutServlet`.
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+To log a user out, you should clear the application session and log the user out of Auth0. This is handled in the `LogoutServlet`.
+
+First, clear the session by calling `request.getSession().invalidate()`. Then construct the logout URL, being sure to include the `returnTo` query parameter, which is where the user will be redirected to after logging out. Finally, redirect the response to the application's logout URL:
```
```suggestion
First, clear the session by calling `request.getSession().invalidate()`. Then construct the logout URL, being sure to include the `returnTo` query parameter, to which the user will be redirected after logging out. Finally, redirect the response to the application's logout URL:
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+## Run the Sample
+
+To build and run the sample, execute the `wildfly:run` Maven goal to start an embedded WildFly application server with this application deployed to it. See the [WildFly Maven Plugin](https://docs.jboss.org/wildfly/plugins/maven/latest/) documentation for more information.
```
```suggestion
To build and run the sample, execute the `wildfly:run` Maven goal. This starts an embedded WildFly application server with the application deployed to it. See the [WildFly Maven Plugin](https://docs.jboss.org/wildfly/plugins/maven/latest/) documentation for more information.
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+## Configure Java EE Security
```
```suggestion
## Configure Java EE security
```

Sentence case for all H2 and smaller

jimmyjames replied:
In this case `Java EE Security` refers to a specific named feature (I think the grammatical term is proper noun?) so I believe "security" should be capitalized.

khorne3 replied:
Gotcha; then let's leave be.

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
```diff
@@ -0,0 +1,379 @@
+
+Finally, note that the `@AutoApplySession` annotation has been added to allow the container to create a session for the authenticated user.
+
+## Trigger Authentication
```
```suggestion
## Trigger authentication
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
````diff
@@ -0,0 +1,379 @@
+}
+```
+
+## Display User Information
````
```suggestion
## Display user information
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
````diff
@@ -0,0 +1,379 @@
+</c:choose>
+```
+
+## Handle Logout
````
```suggestion
## Handle logout
```

khorne3 commented on articles/quickstart/webapp/java-ee/_includes/_login.md:
````diff
@@ -0,0 +1,379 @@
+}
+```
+
+## Run the Sample
````
```suggestion
## Run the sample
```

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_setup.md

Co-Authored-By: K Horne <katie@auth0.com>

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_login.md

Co-Authored-By: K Horne <katie@auth0.com>

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_login.md

Co-Authored-By: K Horne <katie@auth0.com>

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_login.md

Co-Authored-By: K Horne <katie@auth0.com>

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_login.md

Co-Authored-By: K Horne <katie@auth0.com>

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_login.md

Co-Authored-By: K Horne <katie@auth0.com>

jimmyjames committed:
Update articles/quickstart/webapp/java-ee/_includes/_login.md

Co-Authored-By: K Horne <katie@auth0.com>

jimmyjames committed:
Interfaces to implement are not optional

This quickstart requires implementing a few interfaces; clarify language to not suggest that it is optional.
